### DISCUSSION TITLE
Read Only Session for Safety and Readability

### DISCUSSION THREAD

**Discussion by barmanroys** [sqlalchemy/sqlalchemy#11983] [Ideas]
When I create a session to a database, ensure there is a way to mark the session as read-only, i.e. it will *only* query data from the database, without insertion, deletion or upsert on any table. 

Additionally and optionally, it would be great to allow for even more granular permission, such as this session can DELETE from the ORDERS table, no other manipulation allowed. 

Benefits: 
Code readability, safety and conveying intents while creating a session.

**Comment by zzzeek** [comment]
there's nothing out of the box but you would use [SessionEvents](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.SessionEvents) to intercept any activities you want to log or block and raise exceptions within your event handler as needed.     

the most straightforward approach to making sure a session doesnt "write" to the DB is to just prevent a transaction commit from taking place.    but if you want it at the SQL level you want to look at [SessionEvents.do_orm_execute](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.SessionEvents.do_orm_execute) for statements passed to `Session.execute` and for ORM flush events the [SessionEvents.before_flush](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.SessionEvents.before_flush) event, which is only called if a flush detected pending changes to flush.
